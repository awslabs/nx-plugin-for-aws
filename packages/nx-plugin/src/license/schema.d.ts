/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
export type SPDXLicenseIdentifier = 'Apache-2.0' | 'MIT' | 'ASL';

export interface LicenseGeneratorSchema {
  /**
   * SPDX License Identifier
   */
  license: SPDXLicenseIdentifier;

  /**
   * Copyright holder name
   */
  copyrightHolder: string;
}
