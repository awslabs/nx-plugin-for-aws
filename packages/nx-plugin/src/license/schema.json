{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "$id": "License",
  "title": "Add LICENSE files, and configure source code license headers",
  "type": "object",
  "properties": {
    "license": {
      "type": "string",
      "description": "SPDX License Identifier",
      "enum": ["Apache-2.0", "MIT", "ASL"],
      "default": "Apache-2.0",
      "x-priority": "important"
    },
    "copyrightHolder": {
      "type": "string",
      "description": "Copyright holder name",
      "default": "Amazon.com, Inc. or its affiliates",
      "x-priority": "important"
    }
  },
  "required": []
}
