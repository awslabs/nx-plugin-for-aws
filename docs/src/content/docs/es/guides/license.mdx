---
title: "Licencia"
description: "Documentación de referencia para el generador de licencias"
---



import { FileTree } from '@astrojs/starlight/components';
import RunGenerator from '@components/run-generator.astro';
import GeneratorParameters from '@components/generator-parameters.astro';

Administra automáticamente los archivos `LICENSE` y las cabeceras de licencia en el código fuente de tu espacio de trabajo.

Este generador registra un [generador de sincronización](https://nx.dev/concepts/sync-generators) que se ejecuta como parte de tus objetivos `lint`, asegurando que tus archivos fuente cumplan con el contenido y formato de licencia deseado. También garantiza que los archivos `LICENSE` de tus proyectos sean correctos y que la información de licencia esté incluida en archivos de proyecto relevantes (`package.json`, `pyproject.toml`).

## Uso

### Ejecutar el generador

<RunGenerator generator="license" />

### Opciones

<GeneratorParameters generator="license" />

## Salida del generador

El generador creará o actualizará los siguientes archivos:

<FileTree>
  - nx.json El objetivo lint se configura para ejecutar el generador de sincronización de licencias
  - aws-nx-plugin.config.mts Configuración para el generador de sincronización de licencias
</FileTree>

Se agrega una configuración predeterminada para el contenido y formato de cabeceras de licencia en `aws-nx-plugin.config.mts`, adecuada para varios tipos de archivo. Puedes personalizar esto según sea necesario; consulta la [sección de configuración](#configuración) más abajo.

## Flujo de trabajo

Cada vez que construyas tus proyectos (y se ejecute un objetivo `lint`), el generador de sincronización de licencias asegurará que las licencias en tu proyecto coincidan con tu configuración (ver [comportamiento de sincronización más abajo](#comportamiento-de-sincronización-de-licencias)). Si detecta discrepancias, recibirás un mensaje como:

```bash
  NX   El espacio de trabajo está desincronizado

[@aws/nx-plugin:license#sync]: Archivos LICENSE del proyecto están desincronizados:
- LICENSE
- packages/<my-project>LICENSE

Archivos package.json del proyecto están desincronizados:
- package.json

Archivos pyproject.toml del proyecto están desincronizados:
- pyproject.toml
- packages/<my-python-project>/pyproject.toml

Cabeceras de licencia desincronizadas en los siguientes archivos fuente:
- packages/<my-project>/src/index.ts
- packages/<my-python-project>/main.py

Esto resultará en un error en CI.

¿Deseas sincronizar los cambios identificados para actualizar el espacio de trabajo?
Sí, sincronizar los cambios y ejecutar las tareas
No, ejecutar las tareas sin sincronizar los cambios
```

Selecciona `Sí` para sincronizar los cambios.

:::note
Asegúrate de revisar los cambios realizados por el generador en el control de versiones para evitar fallos en tareas de integración continua debido a licencias desincronizadas.
:::

## Comportamiento de sincronización de licencias

El generador realiza tres tareas principales:

### 1. Sincronizar cabeceras de licencia en archivos fuente

El generador asegura que todos los archivos fuente en tu espacio de trabajo (según tu configuración) contengan la cabecera de licencia apropiada. La cabecera se escribe como el primer comentario de bloque o serie de comentarios de línea consecutivos en el archivo (excluyendo shebang/hashbang si está presente).

Puedes actualizar la configuración para cambiar qué archivos incluir/excluir, así como el contenido o formato de cabeceras para diferentes tipos de archivo. Más detalles en la [sección de configuración](#configuración).

### 2. Sincronizar archivos LICENSE

El generador asegura que el archivo raíz `LICENSE` y los archivos `LICENSE` de subproyectos correspondan a tu licencia configurada.

Puedes excluir proyectos en la configuración. Más detalles en la [sección de configuración](#configuración).

### 3. Sincronizar información de licencia en archivos de proyecto

El generador asegura que los campos `license` en archivos `package.json` y `pyproject.toml` coincidan con tu licencia configurada.

Puedes excluir proyectos en la configuración. Más detalles en la [sección de configuración](#configuración).

## Configuración

La configuración se define en el archivo `aws-nx-plugin.config.mts` en la raíz del espacio de trabajo.

### SPDX y Titular de derechos

Puedes actualizar la licencia mediante la propiedad `spdx`:

```typescript title="aws-nx-plugin.config.mts" {3}
export default {
  license: {
    spdx: 'MIT',
  },
} satisfies AwsNxPluginConfig;
```

Al ejecutar el generador, todos los archivos `LICENSE`, `package.json` y `pyproject.toml` se actualizarán según la licencia configurada.

También puedes configurar el titular de derechos y el año de copyright:

```typescript title="aws-nx-plugin.config.mts" {4,5}
export default {
  license: {
    spdx: 'MIT',
    copyrightHolder: 'Amazon.com, Inc. or its affiliates',
    copyrightYear: 2025,
  },
} satisfies AwsNxPluginConfig;
```

### Cabeceras de licencia

#### Contenido

El contenido de la cabecera se puede configurar de dos formas:

1. Contenido inline:

```typescript title="aws-nx-plugin.config.mts" {5-9}
export default {
  license: {
    header: {
      content: {
        lines: [
          'Copyright: My Company, Incorporated.',
          'Licensed under the MIT License',
          'All rights reserved',
        ];
      }
      // ... formato de configuración
    }
  }
} satisfies AwsNxPluginConfig;
```

2. Cargando desde archivo:

```typescript title="aws-nx-plugin.config.mts" {5}
export default {
  license: {
    header: {
      content: {
        filePath: 'license-header.txt'; // relativo a la raíz del workspace
      }
      // ... formato de configuración
    }
  }
} satisfies AwsNxPluginConfig;
```

#### Formato

Puedes especificar el formato de cabeceras para diferentes tipos de archivo usando patrones glob. Soporta comentarios de línea, bloque o combinaciones:

```typescript title="aws-nx-plugin.config.mts" {7-29}
export default {
  license: {
    header: {
      content: {
        lines: ['Aviso de copyright aquí'],
      },
      format: {
        // Comentarios de línea
        '**/*.ts': {
          lineStart: '// ',
        },
        // Comentarios de bloque
        '**/*.css': {
          blockStart: '/*',
          blockEnd: '*/',
        },
        // Comentarios de bloque con prefijos
        '**/*.java': {
          blockStart: '/*',
          lineStart: ' * ',
          blockEnd: ' */',
        },
        // Comentarios con encabezado/pie
        '**/*.py': {
          blockStart: '# ------------',
          lineStart: '# ',
          blockEnd: '# ------------',
        },
      },
    },
  },
} satisfies AwsNxPluginConfig;
```

Opciones de formato soportadas:
- `blockStart`: Texto antes del contenido de licencia
- `lineStart`: Prefijo para cada línea
- `lineEnd`: Sufijo para cada línea
- `blockEnd`: Texto después del contenido

#### Sintaxis de comentarios personalizada

Para tipos de archivo no soportados nativamente:

```typescript title="aws-nx-plugin.config.mts" {12-22}
export default {
  license: {
    header: {
      content: {
        lines: ['Mi cabecera de licencia'],
      },
      format: {
        '**/*.xyz': {
          lineStart: '## ',
        },
      },
      commentSyntax: {
        xyz: {
          line: '##', // Sintaxis de comentario de línea
        },
        abc: {
          block: {
            start: '<!--',
            end: '-->',
          },
        },
      },
    },
  },
} satisfies AwsNxPluginConfig;
```

#### Excluir archivos

Por defecto se respetan los `.gitignore`. En repositorios no-git, se excluyen mediante configuración:

```typescript title="aws-nx-plugin.config.mts" {12-16}
export default {
  license: {
    header: {
      content: {
        lines: ['Mi cabecera de licencia'],
      },
      format: {
        '**/*.ts': {
          lineStart: '// ',
        },
      },
      exclude: ['**/generated/**', '**/dist/**', 'some-specific-file.ts'],
    },
  },
} satisfies AwsNxPluginConfig;
```

### Excluir archivos de proyecto de sincronización

Puedes excluir proyectos o archivos específicos:

```typescript title="aws-nx-plugin.config.mts" {3-10}
export default {
  license: {
    files: {
      exclude: [
        // Excluir LICENSE, package.json y pyproject.toml
        'packages/excluded-project',
        // Excluir solo LICENSE
        'apps/internal/LICENSE',
      ];
    }
  }
} satisfies AwsNxPluginConfig;
```

## Desactivar sincronización de licencias

Para desactivar el generador:
1. Elimina la sección `license` en `aws-nx-plugin.config.mts`
2. Elimina `@aws/nx-plugin:license#sync` de `targetDefaults.lint.syncGenerators`

Para reactivarlo, ejecuta el generador `license` nuevamente.