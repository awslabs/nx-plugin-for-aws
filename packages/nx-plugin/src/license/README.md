# License Generator

## Overview

This generator configures `LICENSE` files and source file headers for your project. After you run this generator, a [sync generator](https://nx.dev/concepts/sync-generators) is registered to execute as part of your `lint` targets which will ensure that your source files conform to the desired license content and format, as well as ensuring that your project's LICENSE files are correct, and licensing information is included in relevant project files.

## Usage

You can run the generator in two ways:

### 1. Using VSCode IDE

First, install the NX Console extension for VSCode:

1. Open VSCode
2. Go to Extensions (Ctrl+Shift+X / Cmd+Shift+X)
3. Search for "Nx Console"
4. Install [Nx Console](https://marketplace.visualstudio.com/items?itemName=nrwl.angular-console)

Then generate your API:

1. Open the NX Console in VSCode
2. Click on "Generate"
3. Search for "license"
4. Fill in the required parameters in the form
5. Click "Run"

### 2. Using CLI

Generate the API:

```bash
nx g @aws/nx-plugin:license my-api --copyrightHolder="My Company, Inc." --license=MIT
```

You can also perform a dry-run to see what files would be generated or updated without actually creating them:

```bash
nx g @aws/nx-plugin:license my-api --copyrightHolder="My Company, Inc." --license=MIT --dry-run
```

Both methods will configure the license sync generator to ensure that your project contains the correct LICENSE files, and source files specify the correct license header.

## Input Parameters

| Parameter       | Type   | Default                              | Description                                                                            |
| --------------- | ------ | ------------------------------------ | -------------------------------------------------------------------------------------- |
| license         | string | "Apache-2.0"                         | The SPDX license identifier for your chosen license.                                   |
| copyrightHolder | string | "Amazon.com, Inc. or its affiliates" | The copyright holder, included in the LICENSE file and source file headers by default. |

## Expected Output

The generator will create or update the following files:

```
└── nx.json                    # The "lint" target is configured to sync LICENSE files and source file headers
└── aws-nx-plugin.config.mts   # Configuration for the license sync, such as customising the license header content and format for different languages
```

Some default configuration for license header content and format is added to `aws-nx-plugin.config.mts` to write appropriate headers for a handful of file types. You may wish to customise this further; please see the [configuration section](#configuration) below.

## License Sync Behaviour

The license sync generator performs three main tasks:

### 1. Synchronise Source File License Headers

When the sync generator is run, it will ensure that all source code files in your workspace (based on your configuration) contain the appropriate license header. The header is written as the first block comment or consecutive series of line comments in the file (besides the shebang/hashbang if present in a file).

You can update the configuration at any time to change which files should be included or excluded, as well as the content or format of license headers for different file types. For more details, please see the [configuration section](#configuration) below.

### 2. Synchronise LICENSE Files

When the sync generator is run, it will ensure that the root `LICENSE` file corresponds to your configured license, as well as ensuring that all subprojects in your workspace also contain the correct `LICENSE` file.

You can exclude projects in the configuration if required. For more details, please see the [configuration section](#configuration) below.

### 3. Synchronise licensing information in project files

When the sync generator is run, it will ensure the `license` fields in `package.json` and `pyproject.toml` files are set to your configured license.

You can exclude projects in the configuration if required. For more details, please see the [configuration section](#configuration) below.

## Configuration

Configuration is defined in the `aws-nx-plugin.config.mts` file in the root of your workspace.

### SPDX and Copyright Holder

Your chosen license can be updated at any time via the `spdx` configuration property:

```typescript
{
  license: {
    spdx: 'MIT',
  }
}
```

When the sync generator runs, all `LICENSE` files, `package.json` and `pyproject.toml` files will be updated to reflect the configured license.

You can additionally configure the copyright holder and copyright year, which are included in some `LICENSE` files:

```typescript
{
  license: {
    spdx: 'MIT',
    copyrightHolder: 'Amazon.com, Inc. or its affiliates',
    copyrightYear: 2025,
  }
}
```

### License Headers

#### Content

The license header content can be configured in two ways:

1. Using inline content:

```typescript
{
  license: {
    header: {
      content: {
        lines: ['Copyright My Company, Inc.', 'Licensed under MIT License', 'All rights reserved'];
      }
      // ... format configuration
    }
  }
}
```

2. Loading from a file:

```typescript
{
  license: {
    header: {
      content: {
        filePath: 'license-header.txt'; // relative to workspace root
      }
      // ... format configuration
    }
  }
}
```

#### Including files and specifying header format

You can specify how license headers should be formatted for different file types using glob patterns. The format configuration supports line comments, block comments, or a combination of both:

```typescript
{
  license: {
    header: {
      content: {
        lines: ['Copyright notice here']
      },
      format: {
        // Line comments
        '**/*.ts': {
          lineStart: '// '
        },

        // Block comments
        '**/*.css': {
          blockStart: '/*',
          blockEnd: '*/'
        },

        // Block comments with line prefixes
        '**/*.java': {
          blockStart: '/*',
          lineStart: ' * ',
          blockEnd: ' */'
        },

        // Line comments with header/footer
        '**/*.py': {
          blockStart: '# ------------',
          lineStart: '# ',
          blockEnd: '# ------------'
        }
      }
    }
  }
}
```

The format configuration supports:

- `blockStart`: Text written before the license content (e.g., to start a block comment)
- `lineStart`: Text prepended to each line of the license content
- `lineEnd`: Text appended to each line of the license content
- `blockEnd`: Text written after the license content (e.g., to end a block comment)

#### Custom comment syntax

For file types that aren't natively supported, you can specify custom comment syntax:

```typescript
{
  license: {
    header: {
      content: {
        lines: ['My license header']
      },
      format: {
        '**/*.xyz': {
          lineStart: '## '
        }
      },
      commentSyntax: {
        xyz: {
          line: '##'  // Define line comment syntax
        },
        abc: {
          block: {    // Define block comment syntax
            start: '<!--',
            end: '-->'
          }
        }
      }
    }
  }
}
```

This tells the sync generator how to identify existing license headers in these file types. The `commentSyntax` configuration supports:

- `line`: Characters that start a line comment
- `block`: Characters that start and end a block comment

#### Excluding files

By default, in a git repository, all `.gitignore` files are honored to ensure that only files managed by version control are synchronized. In non-git repositories, all files are considered unless explicitly excluded in configuration.

You can exclude additional files from license header synchronization using glob patterns:

```typescript
{
  license: {
    header: {
      content: {
        lines: ['My license header']
      },
      format: {
        '**/*.ts': {
          lineStart: '// '
        }
      },
      exclude: [
        '**/generated/**',
        '**/dist/**',
        'some-specific-file.ts'
      ]
    }
  }
}
```

### Excluding project files from sync

All `LICENSE` files, `package.json` files and `pyproject.toml` files are synchronised with the configured license by default.

You can exclude specific projects or files from synchronization using glob patterns:

```typescript
{
  license: {
    files: {
      exclude: [
        'packages/excluded-project', // do not sync LICENSE file, package.json or pyproject.toml
        'apps/internal/LICENSE', // do not sync LICENSE file, but sync package.json and/or pyproject.toml
      ];
    }
  }
}
```

## Disabling license sync

To disable the license sync generator:

1. Remove the `license` section from your configuration in `aws-nx-plugin.config.mts` (or remove the `aws-nx-plugin.config.mts` file)
2. Remove the `@aws/nx-plugin:license#sync` generator from `targetDefaults.lint.syncGenerators`

To re-enable license sync, simply run the `license` generator again.
