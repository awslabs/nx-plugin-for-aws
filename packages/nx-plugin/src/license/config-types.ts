/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
export interface LicenseConfig {
  /**
   * Configuration for source code license headers
   * Omit if you do not want any source code license headers to be synced
   */
  header?: LicenseHeaderConfig;
  /**
   * Configuration for LICENSE files
   */
  files?: LicenseFileConfig;
}

/**
 * Configuration for the LICENSE file sync
 */
export interface LicenseFileConfig {
  /**
   * Exclude LICENSE file sync checks for the given glob patterns
   */
  exclude?: string[];
}

/**
 * Specifies license content as a path to a file
 */
export interface LicenseFilePathContent {
  /**
   * Load the license text from the specified file (relative to the workspace root)
   */
  filePath: string;
}

/**
 * Specifies license content inline
 */
export interface LicenseLinesContent {
  /**
   * Inline license text
   */
  lines: string[];
}

/**
 * Configuration for source code license header sync
 */
export interface LicenseHeaderConfig {
  /**
   * Content of the license header
   */
  content: LicenseFilePathContent | LicenseLinesContent;
  /**
   * Specifies the license header format for given glob patterns
   */
  format: {
    [filePattern: string]: LicenseHeaderFormat;
  };
  /**
   * Exclude out of sync source code license header checks for the given glob patterns
   */
  exclude?: string[];
  /**
   * User-specified comment syntax for file types we don't have config for
   */
  commentSyntax?: {
    [ext: string]: CommentSyntax;
  };
}

/**
 * Format settings for source code license headers
 */
export interface LicenseHeaderFormat {
  /**
   * Text written to the line prior to the license lines
   * Use this to start a block comment if desired
   */
  blockStart?: string;
  /**
   * Text written to the start of each line of the license
   */
  lineStart?: string;
  /**
   * Text written to the end of each line of the license
   */
  lineEnd?: string;
  /**
   * Text written to the line after the license lines
   * Use this to end a block comment if desired
   */
  blockEnd?: string;
}

/**
 * Defines comment syntax for a particular language
 * Used to instruct the license sync generator to find the first comment in a file
 */
export interface CommentSyntax {
  /**
   * If the language supports line comments, specify the character(s) used to start a line comment
   */
  line?: string;
  /**
   * If the language supports block comments, specify the character(s) used to start and end the block comment
   */
  block?: {
    start: string;
    end: string;
  };
}
