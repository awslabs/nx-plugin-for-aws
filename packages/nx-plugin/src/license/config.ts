/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree } from '@nx/devkit';
import { SPDXLicenseIdentifier } from './schema';
import {
  readAwsNxPluginConfig,
  updateAwsNxPluginConfig,
} from '../utils/config/utils';
import { CommentSyntax, LicenseConfig } from './config-types';

/**
 * Licenses containing a placeholder for the copyright year
 */
const LICENSES_WITH_YEAR: Set<SPDXLicenseIdentifier> = new Set(['MIT']);

/**
 * Defines the comment syntax for popular programming languages
 */
export const LANGUAGE_COMMENT_SYNTAX: { [ext: string]: CommentSyntax } = {
  // C++ style comments
  ...Object.fromEntries(
    [
      // Node
      'js',
      'ts',
      'cjs',
      'mjs',
      'cts',
      'mts',
      // React
      'jsx',
      'tsx',
      // Web
      'sass',
      'scss',
      'php',
      // Java
      'java',
      // Smithy
      'smithy',
      // TypeSpec
      'tsp',
      // Golang
      'go',
      // Rust
      'rs',
      // C and C++
      'c',
      'cpp',
      'cxx',
      'cc',
      'h',
      'hpp',
      // C#
      'cs',
      // Scala
      'scala',
      'sc',
      'sbt',
      // Kotlin
      'kt',
      // Swift
      'swift',
    ].map((ext) => [ext, { line: '//', block: { start: '/*', end: '*/' } }]),
  ),
  // Bash style comments
  ...Object.fromEntries(
    [
      // Bash
      'sh',
      // Perl
      'pl',
      // YAML
      'yaml',
      'yml',
      // R
      'r',
    ].map((ext) => [ext, { line: '#' }]),
  ),
  // Docker
  Dockerfile: { line: '#' },
  // Python
  py: { line: '#', block: { start: '"""', end: '"""' } },
  // Web
  html: { block: { start: '<!--', end: '-->' } },
  css: { block: { start: '/*', end: '*/' } },
  // Ruby
  rb: { line: '#', block: { start: '=begin', end: '=end' } },
  // F#
  ...Object.fromEntries(
    ['fs', 'fsx'].map((ext) => [
      ext,
      { line: '//', block: { start: '(*', end: '*)' } },
    ]),
  ),
  // Visual Basic
  vb: { line: "' " },
  // Lua
  lua: { line: '--', block: { start: '--[[', end: ']]' } },
  // PowerShell
  ps1: { line: '#', block: { start: '<#', end: '#>' } },
  psm1: { line: '#', block: { start: '<#', end: '#>' } },
  // Markdown
  md: { block: { start: '<!--', end: '-->' } },
};

/**
 * Build the default license config for a given license
 */
export const defaultLicenseConfig = (
  spdx: SPDXLicenseIdentifier,
  copyrightHolder: string,
): LicenseConfig => {
  switch (spdx) {
    case 'ASL': {
      // For ASL, we use a "box" style license header
      const rawContent = [
        `Copyright ${copyrightHolder}. All Rights Reserved.`,
        '',
        'Licensed under the Amazon Software License (the "License"). You may not use this file except in compliance',
        'with the License. A copy of the License is located at',
        '',
        '   https://aws.amazon.com/asl/',
        '',
        'or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES',
        'OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions',
        'and limitations under the License.',
      ];
      const maxLen = Math.max(...rawContent.map((line) => line.length));
      const lines = rawContent.map(
        (line) => `${line}${' '.repeat(maxLen - line.length)}`,
      );
      return {
        spdx,
        copyrightHolder,
        header: {
          content: { lines },
          format: {
            '**/*.{js,ts,jsx,tsx,mjs,mts}': {
              blockStart: `/***${'*'.repeat(maxLen)}**`,
              lineStart: ' *  ',
              lineEnd: ' *',
              blockEnd: ` ***${'*'.repeat(maxLen)}**/`,
            },
            '**/*.{py,sh}': {
              blockStart: `###${'#'.repeat(maxLen)}##`,
              lineStart: '#  ',
              lineEnd: ' #',
              blockEnd: `###${'#'.repeat(maxLen)}##`,
            },
          },
          exclude: ['**/*.gen.*'],
        },
      };
    }
    default: {
      return {
        spdx,
        copyrightHolder,
        ...(LICENSES_WITH_YEAR.has(spdx)
          ? {
              copyrightYear: new Date().getFullYear(),
            }
          : {}),
        header: {
          content: {
            lines: [
              `Copyright ${copyrightHolder}. All Rights Reserved.`,
              `SPDX-License-Identifier: ${spdx}`,
            ],
          },
          format: {
            '**/*.{js,ts,jsx,tsx,mjs,mts}': {
              blockStart: '/**',
              lineStart: ' * ',
              blockEnd: ' */',
            },
            '**/*.{py,sh}': {
              blockStart: '#',
              lineStart: '# ',
              blockEnd: '#',
            },
          },
          exclude: ['**/*.gen.*'],
        },
      };
    }
  }
};

/**
 * Write license configuration to the aws nx plugin config
 */
export const writeLicenseConfig = async (tree: Tree, config: LicenseConfig) => {
  await updateAwsNxPluginConfig(tree, {
    license: config,
  });
};

/**
 * Read license configuration
 */
export const readLicenseConfig = async (
  tree: Tree,
): Promise<LicenseConfig | undefined> => {
  return (await readAwsNxPluginConfig(tree))?.license;
};
