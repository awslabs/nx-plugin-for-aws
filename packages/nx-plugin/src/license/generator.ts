/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { generateFiles, readNxJson, Tree, updateNxJson } from '@nx/devkit';
import { join } from 'path';
import { LicenseGeneratorSchema } from './schema';
import { defaultLicenseConfig, writeLicenseConfig } from './config';
import { ensureAwsNxPluginConfig } from '../utils/config/utils';
import { SYNC_GENERATOR_NAME } from './sync/generator';

export async function licenseGenerator(
  tree: Tree,
  options: LicenseGeneratorSchema,
) {
  const { license, copyrightHolder } = options;

  // Add LICENSE file
  generateFiles(tree, join(__dirname, 'files', 'licenses', license), '.', {
    ...options,
    year: new Date().getFullYear(),
  });

  // Write default config for the license headers
  await ensureAwsNxPluginConfig(tree);
  await writeLicenseConfig(
    tree,
    license,
    defaultLicenseConfig(license, copyrightHolder),
  );

  // Configure sync generator to run as part of all projects' lint target
  const nxJson = readNxJson(tree);
  updateNxJson(tree, {
    ...nxJson,
    targetDefaults: {
      ...nxJson.targetDefaults,
      lint: {
        ...nxJson.targetDefaults?.lint,
        syncGenerators: [
          ...(nxJson.targetDefaults?.lint?.syncGenerators ?? []).filter(
            (g) => g !== SYNC_GENERATOR_NAME,
          ),
          SYNC_GENERATOR_NAME,
        ],
      },
    },
  });
}

export default licenseGenerator;
