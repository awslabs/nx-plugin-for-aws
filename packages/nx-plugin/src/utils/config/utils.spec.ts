/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { Tree } from '@nx/devkit';
import {
  ensureAwsNxPluginConfig,
  readAwsNxPluginConfig,
  updateAwsNxPluginConfig,
  AWS_NX_PLUGIN_CONFIG_FILE_NAME,
} from './utils';
import { LicenseLinesContent } from '../../license/config-types';

describe('config utils', () => {
  let tree: Tree;

  beforeEach(async () => {
    tree = createTreeWithEmptyWorkspace();
    await ensureAwsNxPluginConfig(tree);
  });

  it('should ensure config exists', async () => {
    expect(tree.exists(AWS_NX_PLUGIN_CONFIG_FILE_NAME)).toBe(true);
  });

  it('should update user config', async () => {
    await updateAwsNxPluginConfig(tree, {
      license: {
        header: {
          content: { lines: ['Test Copyright Header'] },
          format: {
            '**/*.ts': {
              blockStart: '/**',
              lineStart: ' * ',
              blockEnd: ' */',
            },
          },
        },
      },
    });

    expect(
      tree.read(AWS_NX_PLUGIN_CONFIG_FILE_NAME, 'utf-8'),
    ).toMatchSnapshot();

    expect(
      (
        (await readAwsNxPluginConfig(tree)).license.header
          .content as LicenseLinesContent
      ).lines[0],
    ).toBe('Test Copyright Header');

    await updateAwsNxPluginConfig(tree, {
      license: {
        header: {
          content: { lines: ['Test Copyright Header 2'] },
          format: {
            '**/*.ts': {
              blockStart: '/**',
              lineStart: ' * ',
              blockEnd: ' */',
            },
          },
        },
      },
    });

    expect(
      (
        (await readAwsNxPluginConfig(tree)).license.header
          .content as LicenseLinesContent
      ).lines[0],
    ).toBe('Test Copyright Header 2');
  });

  it('should not delete other user config', async () => {
    await updateAwsNxPluginConfig(tree, {
      license: {
        header: {
          content: { lines: ['Test Copyright Header'] },
          format: {
            '**/*.ts': {
              blockStart: '/**',
              lineStart: ' * ',
              blockEnd: ' */',
            },
          },
        },
      },
    });

    await updateAwsNxPluginConfig(tree, {});

    expect(
      (
        (await readAwsNxPluginConfig(tree)).license.header
          .content as LicenseLinesContent
      ).lines[0],
    ).toBe('Test Copyright Header');
  });
});
