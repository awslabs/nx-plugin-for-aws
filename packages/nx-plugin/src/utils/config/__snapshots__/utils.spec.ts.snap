// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`config utils > should update user config 1`] = `
"import { AwsNxPluginConfig } from '@aws/nx-plugin';

export default {
  license: {
    header: {
      content: { lines: ['Test Copyright Header'] },
      format: {
        '**/*.ts': { blockStart: '/**', lineStart: ' * ', blockEnd: ' */' },
      },
    },
  },
} satisfies AwsNxPluginConfig;
"
`;
